File: apps/web/src/features/settings/api/memories.ts
import type {
  MemoryEditRequest,
  MemoryItem,
  MemoryKind,
  MemoryListResponse,
  MemorySearchResponse,
} from "@pathfinder/shared";

import { getAuthHeaders } from "@/lib/api/http";

const BASE = "/api/v1/memories";

export async function listMemories(opts?: {
  limit?: number;
  offset?: number;
}): Promise<MemoryListResponse> {
  const params = new URLSearchParams();
  if (opts?.limit != null) params.set("limit", String(opts.limit));
  if (opts?.offset != null) params.set("offset", String(opts.offset));
  const url = params.size > 0 ? `${BASE}?${params.toString()}` : BASE;
  const res = await fetch(url, {
    credentials: "include",
    headers: getAuthHeaders(),
  });
  if (!res.ok) throw new Error(`listMemories: ${res.status}`);
  return (await res.json()) as MemoryListResponse;
}

export async function searchMemories(
  query: string,
): Promise<MemorySearchResponse> {
  const res = await fetch(
    `${BASE}/search?q=${encodeURIComponent(query)}`,
    { credentials: "include", headers: getAuthHeaders() },
  );
  if (!res.ok) throw new Error(`searchMemories: ${res.status}`);
  return (await res.json()) as MemorySearchResponse;
}

export async function editMemory(
  key: string,
  kind: MemoryKind,
  body: MemoryEditRequest,
): Promise<MemoryItem> {
  const res = await fetch(
    `${BASE}/${encodeURIComponent(key)}?kind=${encodeURIComponent(kind)}`,
    {
      method: "PATCH",
      credentials: "include",
      headers: getAuthHeaders({ contentType: "application/json" }),
      body: JSON.stringify(body),
    },
  );
  if (!res.ok) throw new Error(`editMemory: ${res.status}`);
  return (await res.json()) as MemoryItem;
}

export async function deleteMemory(
  key: string,
  kind: MemoryKind,
): Promise<void> {
  const res = await fetch(
    `${BASE}/${encodeURIComponent(key)}?kind=${encodeURIComponent(kind)}`,
    {
      method: "DELETE",
      credentials: "include",
      headers: getAuthHeaders(),
    },
  );
  if (!res.ok && res.status !== 204) {
    throw new Error(`deleteMemory: ${res.status}`);
  }
}

File: apps/api/src/pathfinder/transport/http/routers/memories.py
from __future__ import annotations

from typing import Annotated

from assistant_core.memory.schemas import MemoryKind
from assistant_core.memory.store import MemoryStore, StoredMemory
from assistant_core.memory.tombstones import TombstoneRepository
from assistant_core.platform.db import async_session_factory
from fastapi import APIRouter, HTTPException, Query, Request, Response, status
from langgraph.store.postgres.aio import AsyncPostgresStore

from pathfinder.transport.http.deps import CurrentUser
from pathfinder.transport.http.schemas.memories import (
    MemoryEditRequest,
    MemoryItem,
    MemoryListResponse,
    MemorySearchResponse,
)

router = APIRouter(prefix="/api/v1/memories", tags=["memories"])

# The order the response fields are filled in; the set is pinned against the
# assistant's declared kinds.
MEMORY_ROUTE_KINDS: tuple[MemoryKind, ...] = (
    "gene_set",
    "strategy",
    "preference",
    "knowledge",
)

_MAX_PAGE_LIMIT: int = 200
_DEFAULT_PAGE_LIMIT: int = 50


def _store(request: Request) -> MemoryStore:
    raw: AsyncPostgresStore = request.app.state.memory_store
    return MemoryStore(store=raw)


def _tombstones() -> TombstoneRepository:
    return TombstoneRepository(session_factory=async_session_factory)


def _to_item(stored: StoredMemory) -> MemoryItem:
    return MemoryItem(key=stored.key, value=stored.value)


@router.get("", response_model=MemoryListResponse)
async def list_memories(
    request: Request,
    user_id: CurrentUser,
    limit: Annotated[
        int,
        Query(ge=1, le=_MAX_PAGE_LIMIT, description="Per-namespace page size"),
    ] = _DEFAULT_PAGE_LIMIT,
    offset: Annotated[int, Query(ge=0, description="Per-namespace offset")] = 0,
) -> MemoryListResponse:
    store = _store(request)
    buckets: dict[MemoryKind, list[MemoryItem]] = {}
    any_full_page = False
    for kind in MEMORY_ROUTE_KINDS:
        stored = await store.list_all(
            user_id=user_id,
            kind=kind,
            limit=limit,
            offset=offset,
        )
        buckets[kind] = [_to_item(s) for s in stored]
        # If any namespace returned a full page, more rows may exist at
        # the next offset — the client should show "load more".
        if len(stored) >= limit:
            any_full_page = True
    return MemoryListResponse(
        gene_sets=buckets["gene_set"],
        strategies=buckets["strategy"],
        preferences=buckets["preference"],
        knowledge=buckets["knowledge"],
        page_size=limit,
        offset=offset,
        has_more=any_full_page,
    )


@router.get("/search", response_model=MemorySearchResponse)
async def search_memories(
    request: Request,
    user_id: CurrentUser,
    q: Annotated[str, Query(min_length=1, max_length=500)],
) -> MemorySearchResponse:
    store = _store(request)
    hits: list[MemoryItem] = []
    for kind in MEMORY_ROUTE_KINDS:
        stored = await store.semantic_search(
            user_id=user_id,
            kind=kind,
            query=q,
            top_k=5,
        )
        hits.extend(_to_item(s) for s in stored)
    return MemorySearchResponse(hits=hits[:10])


@router.patch("/{key}", response_model=MemoryItem)
async def edit_memory(
    key: str,
    kind: Annotated[MemoryKind, Query()],
    body: MemoryEditRequest,
    request: Request,
    user_id: CurrentUser,
) -> MemoryItem:
    store = _store(request)
    current = await store.get(user_id=user_id, kind=kind, key=key)
    if current is None:
        raise HTTPException(status_code=404, detail="memory not found")
    updates = {
        k: v for k, v in body.model_dump(exclude_unset=True).items() if v is not None
    }
    updated = current.value.model_copy(update=updates)
    await store.put(user_id=user_id, value=updated, key=key)
    return MemoryItem(key=key, value=updated)


@router.delete("/{key}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_memory(
    key: str,
    kind: Annotated[MemoryKind, Query()],
    request: Request,
    user_id: CurrentUser,
) -> Response:
    store = _store(request)
    tombstones = _tombstones()
    current = await store.get(user_id=user_id, kind=kind, key=key)
    if current is None:
        raise HTTPException(status_code=404, detail="memory not found")
    await tombstones.tombstone(user_id=user_id, value=current.value)
    await store.delete(user_id=user_id, kind=kind, key=key)
    return Response(status_code=status.HTTP_204_NO_CONTENT)
